server/src/com/thoughtworks/go/server/scheduling/BuildCauseProducerService.java

```diff
@@ -249,6 +258,52 @@ private void updateChangedRevisions(CaseInsensitiveString pipelineName, BuildCau
         materialChecker.updateChangedRevisions(pipelineName, buildCause);
     }
 
+    void pruneRevisionsIfRequired(BuildType buildType, PipelineConfig pipelineConfig, BuildCause buildCause) {
```

Comment:
I don't think this is the right way to do this. This work should not be done here. It should be done right after [this](https://github.com/gocd/gocd/blob/master/server/src/com/thoughtworks/go/server/scheduling/BuildCauseProducerService.java#L184) and [this](https://github.com/gocd/gocd/blob/master/server/src/com/thoughtworks/go/server/scheduling/BuildCauseProducerService.java#L200) and should happen inside [AutoBuild](https://github.com/gocd/gocd/blob/master/server/src/com/thoughtworks/go/server/service/AutoBuild.java#L40). Otherwise, fan-in would have run, inside [AutoBuild#onModifications](https://github.com/gocd/gocd/blob/master/server/src/com/thoughtworks/go/server/service/AutoBuild.java#L59) and would have decided a compatible revision, based on the top-most material revision, which this code happily goes and (possibly) prunes. I think the revisions passed into ```onModifications``` should itself be changed, in the places I mention above, so that checks like ```buildType.shouldCheckWhetherOlderRunsHaveRunWithLatestMaterials()``` etc. would have run against the correct revisions.

I could be wrong, since you actually know more about this area than I do :), however, I am having a very bad reaction to seeing revisions thrown away, just before scheduling. Even if what you've done is correct, consider what I say, about pruning before starting, or just not considering extra revisions before running fan-in.

I think this is a change that we shouldn't try and rush in, in this release. The only part it will affect is that in the GitHub PR, if multiple PRs happen at close to the same time, they will be bunched together, right? That's ok. We will live with this. I'd rather take time, and get it right, since this area is critical, and getting it wrong here could have huge consequences. I want more people to look at it, for instance, @jyotisingh and others.